File: stats-server/src/seed.ts
/**
 * Development script that fills the games database with simulated finished games,
 * so the stats MCP server has data to show.
 *
 * Usage: npm run seed --workspace=stats-server
 */

import path from 'path'
import { fileURLToPath, pathToFileURL } from 'url'
import Database from 'better-sqlite3'
import {
  TicTacToeGame,
  RockPaperScissorsGame,
  ConnectFourGame,
  GAME_TYPES,
  DIFFICULTIES,
  type BaseGameState,
  type Difficulty,
  type Game,
  type GameMove,
  type GameSession,
  type GameType,
  type Player,
  type PlayerId,
  type RPSChoice,
  type RPSGameState
} from '@turn-based-mcp/shared'

type SeatId = 'player1' | 'ai'
type Skills = Record<SeatId, number>

export type SeedSession = GameSession & { gameType: GameType; difficulty: Difficulty }

export interface SeedOptions {
  gamesPerPlayer?: number
  random?: () => number
  now?: Date
}

/** Simulated players and how often they spot a winning or blocking move */
const SEED_PLAYERS = [
  { name: 'Ana', skill: 0.8 },
  { name: 'Bruno', skill: 0.3 },
  { name: 'Caro', skill: 0.6 },
  { name: 'Dani', skill: 0.5 },
]

/** How often the AI spots a winning or blocking move on each difficulty */
const AI_SKILL: Record<Difficulty, number> = { easy: 0.1, medium: 0.6, hard: 0.95 }

const SEED_DAYS = 14
const DAY_MS = 24 * 60 * 60 * 1000

const RPS_CHOICES: RPSChoice[] = ['rock', 'paper', 'scissors']
/** The choice that beats each key */
const BEATEN_BY: Record<RPSChoice, RPSChoice> = { rock: 'paper', paper: 'scissors', scissors: 'rock' }

const TABLES: Record<GameType, string> = {
  'tic-tac-toe': 'tic_tac_toe_games',
  'rock-paper-scissors': 'rps_games',
  'connect-four': 'connect_four_games',
}

const DEFAULT_DB_PATH = path.resolve(path.dirname(fileURLToPath(import.meta.url)), '../../web/games.db')

/**
 * Small deterministic pseudo-random generator (mulberry32), so a seed can be reproduced
 */
export function createRandom(seed: number): () => number {
  let state = seed >>> 0
  return (): number => {
    state = (state + 0x6d2b79f5) >>> 0
    let t = state
    t = Math.imul(t ^ (t >>> 15), t | 1)
    t ^= t + Math.imul(t ^ (t >>> 7), t | 61)
    return ((t ^ (t >>> 14)) >>> 0) / 4294967296
  }
}

/**
 * Plays complete games between the seed players and the AI using the real game rules.
 * Game types rotate so every type is represented; difficulty and dates are random.
 */
export function simulateSessions({ gamesPerPlayer = 10, random = Math.random, now = new Date() }: SeedOptions = {}): SeedSession[] {
  const sessions: SeedSession[] = []

  for (const seedPlayer of SEED_PLAYERS) {
    for (let i = 0; i < gamesPerPlayer; i++) {
      const gameType = GAME_TYPES[i % GAME_TYPES.length]
      const difficulty = DIFFICULTIES[Math.floor(random() * DIFFICULTIES.length)]
      const players: Player[] = [
        { id: 'player1', name: seedPlayer.name, isAI: false },
        { id: 'ai', name: 'AI', isAI: true },
      ]
      const skills: Skills = { player1: seedPlayer.skill, ai: AI_SKILL[difficulty] }

      const { gameState, history } = gameType === 'rock-paper-scissors'
        ? playRockPaperScissors(players, skills, random)
        : gameType === 'tic-tac-toe'
          ? playBoardGame(new TicTacToeGame(), players, skills, random)
          : playBoardGame(new ConnectFourGame(), players, skills, random)

      const finishedAt = new Date(now.getTime() - random() * SEED_DAYS * DAY_MS)
      gameState.createdAt = new Date(finishedAt.getTime() - (2 + random() * 8) * 60 * 1000)
      gameState.updatedAt = finishedAt

      sessions.push({ gameState, gameType, history, difficulty })
    }
  }

  return sessions
}

function playBoardGame<TState extends BaseGameState, TMove>(
  game: Game<TState, TMove>,
  players: Player[],
  skills: Skills,
  random: () => number
): { gameState: TState; history: GameMove[] } {
  let state = game.getInitialState(players, { firstPlayerId: random() < 0.5 ? 'player1' : 'ai' })
  const history: GameMove[] = []
  let result = game.checkGameEnd(state)

  while (!result) {
    const mover = state.currentPlayerId as SeatId
    const move = chooseBoardMove(game, state, mover, skills[mover], random)
    state = game.applyMove(state, move, mover)
    history.push({ playerId: mover, move, timestamp: new Date() })
    result = game.checkGameEnd(state)
  }

  return { gameState: { ...state, status: 'finished', winner: result.winner }, history }
}

/**
 * A skilled mover wins when it can, otherwise blocks the opponent; everyone else plays randomly
 */
function chooseBoardMove<TState extends BaseGameState, TMove>(
  game: Game<TState, TMove>,
  state: TState,
  mover: SeatId,
  skill: number,
  random: () => number
): TMove {
  if (random() < skill) {
    const opponent: SeatId = mover === 'ai' ? 'player1' : 'ai'
    const smartMove = findWinningMove(game, state, mover)
      ?? findWinningMove(game, { ...state, currentPlayerId: opponent }, opponent)
    if (smartMove !== undefined) {
      return smartMove
    }
  }

  const validMoves = game.getValidMoves(state, mover)
  return validMoves[Math.floor(random() * validMoves.length)]
}

function findWinningMove<TState extends BaseGameState, TMove>(game: Game<TState, TMove>, state: TState, playerId: PlayerId): TMove | undefined {
  return game.getValidMoves(state, playerId)
    .find(move => game.checkGameEnd(game.applyMove(state, move, playerId))?.winner === playerId)
}

/**
 * Best of 3. A skilled side sometimes "reads" the other choice and counters it.
 */
function playRockPaperScissors(
  players: Player[],
  skills: Skills,
  random: () => number
): { gameState: RPSGameState; history: GameMove[] } {
  const game = new RockPaperScissorsGame()
  let state = game.getInitialState(players, { maxRounds: 3 })
  const history: GameMove[] = []
  let result = game.checkGameEnd(state)

  while (!result) {
    let playerChoice = RPS_CHOICES[Math.floor(random() * RPS_CHOICES.length)]
    let aiChoice = RPS_CHOICES[Math.floor(random() * RPS_CHOICES.length)]
    if (random() < skills.ai * 0.4) {
      aiChoice = BEATEN_BY[playerChoice]
    } else if (random() < skills.player1 * 0.4) {
      playerChoice = BEATEN_BY[aiChoice]
    }

    state = game.applyMove(state, { choice: playerChoice }, 'player1')
    state = game.applyMove(state, { choice: aiChoice }, 'ai')
    history.push(
      { playerId: 'player1', move: { choice: playerChoice }, timestamp: new Date() },
      { playerId: 'ai', move: { choice: aiChoice }, timestamp: new Date() }
    )
    result = game.checkGameEnd(state)
  }

  return { gameState: { ...state, status: 'finished', winner: result.winner }, history }
}

function toSqlDate(date: Date): string {
  return new Date(date).toISOString().replace('T', ' ').slice(0, 19)
}

function writeSessions(dbPath: string, sessions: SeedSession[]): void {
  const db = new Database(dbPath)

  for (const table of Object.values(TABLES)) {
    db.exec(`
      CREATE TABLE IF NOT EXISTS ${table} (
        id TEXT PRIMARY KEY,
        game_session TEXT NOT NULL,
        created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
        updated_at DATETIME DEFAULT CURRENT_TIMESTAMP
      )
    `)
  }

  const insertAll = db.transaction((items: SeedSession[]) => {
    for (const session of items) {
      db.prepare(`INSERT OR REPLACE INTO ${TABLES[session.gameType]} (id, game_session, created_at, updated_at) VALUES (?, ?, ?, ?)`)
        .run(session.gameState.id, JSON.stringify(session), toSqlDate(session.gameState.createdAt), toSqlDate(session.gameState.updatedAt))
    }
  })
  insertAll(sessions)

  db.close()
}

function main(): void {
  const dbPath = process.env.GAMES_DB_PATH || DEFAULT_DB_PATH
  const sessions = simulateSessions({ gamesPerPlayer: 10 })

  writeSessions(dbPath, sessions)
  console.log(`Seeded ${sessions.length} finished games for ${SEED_PLAYERS.map(p => p.name).join(', ')} into ${dbPath}`)
}

// Only run when executed directly (node dist/seed.js), not when imported by tests
if (process.argv[1] && import.meta.url === pathToFileURL(process.argv[1]).href) {
  main()
}

File: stats-mcp/seed.py
"""Fills the games database with simulated finished games so the stats server has data to show.

Usage (from the repo root): uv run --directory stats-mcp python seed.py
Writes to web/games.db unless GAMES_DB_PATH points somewhere else. Every run adds 40 new games.
"""

import json
import os
import random
import sqlite3
import uuid
from datetime import datetime, timedelta, timezone
from pathlib import Path

DEFAULT_DB_PATH = Path(__file__).resolve().parent.parent / "web" / "games.db"

TABLES = {
    "tic-tac-toe": "tic_tac_toe_games",
    "rock-paper-scissors": "rps_games",
    "connect-four": "connect_four_games",
}
DIFFICULTIES = ["easy", "medium", "hard"]

# Base chance of each simulated player beating the AI
PLAYER_SKILL = {"Ana": 0.65, "Bruno": 0.35, "Caro": 0.55, "Dani": 0.45}
# Added to the player's chance on each difficulty
DIFFICULTY_ADJUSTMENT = {"easy": 0.15, "medium": -0.1, "hard": -0.3}
DRAW_CHANCE = 0.12

# Best of 3 final scores (player-ai) that match each winner
RPS_SCORES = {"player1": [(2, 0), (2, 1)], "ai": [(0, 2), (1, 2)], "draw": [(1, 1), (0, 0)]}

GAMES_PER_PLAYER = 10
SEED_DAYS = 14


def to_iso(moment: datetime) -> str:
    """Same format the web app stores (JavaScript Date JSON): 2026-09-10T16:51:45.258Z"""
    return moment.astimezone(timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")


def pick_winner(skill: float, difficulty: str) -> str:
    if random.random() < DRAW_CHANCE:
        return "draw"
    return "player1" if random.random() < skill + DIFFICULTY_ADJUSTMENT[difficulty] else "ai"


def build_session(player: str, game_type: str, difficulty: str, winner: str, finished_at: datetime) -> dict:
    game_state = {
        "id": str(uuid.uuid4()),
        "players": [
            {"id": "player1", "name": player, "isAI": False},
            {"id": "ai", "name": "AI", "isAI": True},
        ],
        "currentPlayerId": "player1",
        "status": "finished",
        "winner": winner,
        "createdAt": to_iso(finished_at - timedelta(minutes=random.randint(2, 10))),
        "updatedAt": to_iso(finished_at),
    }

    # Minimal game-specific fields so the web app and the games MCP can still open these games
    if game_type == "tic-tac-toe":
        game_state["board"] = [[None] * 3 for _ in range(3)]
        game_state["playerSymbols"] = {"player1": "X", "ai": "O"}
    elif game_type == "connect-four":
        game_state["board"] = [[None] * 7 for _ in range(6)]
        game_state["playerDiscs"] = {"player1": "R", "ai": "Y"}
    else:
        player_score, ai_score = random.choice(RPS_SCORES[winner])
        game_state["rounds"] = [{}, {}, {}]
        game_state["currentRound"] = 3
        game_state["maxRounds"] = 3
        game_state["scores"] = {"player1": player_score, "ai": ai_score}

    return {"gameState": game_state, "gameType": game_type, "history": [], "difficulty": difficulty}


def main() -> None:
    db_path = Path(os.environ.get("GAMES_DB_PATH") or DEFAULT_DB_PATH)
    now = datetime.now(timezone.utc)
    game_types = list(TABLES)

    sessions = []
    for player, skill in PLAYER_SKILL.items():
        for i in range(GAMES_PER_PLAYER):
            game_type = game_types[i % len(game_types)]
            difficulty = random.choice(DIFFICULTIES)
            winner = pick_winner(skill, difficulty)
            finished_at = now - timedelta(seconds=random.uniform(0, SEED_DAYS * 24 * 3600))
            sessions.append(build_session(player, game_type, difficulty, winner, finished_at))

    db = sqlite3.connect(db_path)
    with db:
        for table in TABLES.values():
            db.execute(
                f"""CREATE TABLE IF NOT EXISTS {table} (
                    id TEXT PRIMARY KEY,
                    game_session TEXT NOT NULL,
                    created_at DATETIME DEFAULT CURRENT_TIMESTAMP,
                    updated_at DATETIME DEFAULT CURRENT_TIMESTAMP
                )"""
            )
        for session in sessions:
            state = session["gameState"]
            db.execute(
                f"INSERT OR REPLACE INTO {TABLES[session['gameType']]} (id, game_session, created_at, updated_at) VALUES (?, ?, ?, ?)",
                (state["id"], json.dumps(session), state["createdAt"][:19].replace("T", " "), state["updatedAt"][:19].replace("T", " ")),
            )
    db.close()

    print(f"Seeded {len(sessions)} finished games for {', '.join(PLAYER_SKILL)} into {db_path}")


if __name__ == "__main__":
    main()
